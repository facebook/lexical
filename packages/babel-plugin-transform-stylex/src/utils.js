/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

'use strict';

const messages = require('./messages.js');
const path = require('path');

/**
 * Dashify a camel cased string. This convers the JS property keys to CSS keys.
 */
function dashify(str) {
  return str.replace(/(^|[a-z])([A-Z])/g, '$1-$2').toLowerCase();
}

// Validate an individual key in the object passed to
// stylex.create. The key must be a static value.
function validateProp(prop) {
  // Don't allow computed properties or object spread
  if (!prop.isObjectProperty()) {
    throw prop.buildCodeFrameError(messages.NON_STATIC_VALUE);
  }
  if (
    prop.is('computed') &&
    !prop.get('key').isStringLiteral() &&
    !prop.get('key').isNumericLiteral()
  ) {
    throw prop.buildCodeFrameError(messages.NON_STATIC_VALUE);
  }
}

function getKey(key) {
  if (key.isIdentifier()) {
    return key.node.name;
  } else if (key.isStringLiteral() || key.isNumericLiteral()) {
    return key.node.value;
  } else {
    throw key.buildCodeFrameError(messages.UNKNOWN_PROP_KEY);
  }
}

/**
 * This method is used to generate a development class name from a namespace
 * and a filename. This is only inserted in development and is just a pretty
 * name so you can easily map a DOM element instead of having to use the React
 * inspector.
 */
function namespaceToDevClassName(namespace, filename) {
  // Get the basename of the file without the extension
  const basename = path.basename(filename).split('.')[0];

  // Build up the class name, and sanitize it of disallowed characters
  const className = `${basename}__${namespace}`;
  return className.replace(/[^.a-zA-Z0-9_-]/g, '');
}

module.exports = {
  dashify,
  getKey,
  namespaceToDevClassName,
  validateProp,
};
