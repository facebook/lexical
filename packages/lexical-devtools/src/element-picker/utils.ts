/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

export interface BoundingBox {
  x: number;
  y: number;
  width: number;
  height: number;
}

export interface ElementOverlayStyleOptions {
  background?: string;
  borderColor?: string;
  borderStyle?: string;
  borderRadius?: string;
  borderWidth?: string;
  boxSizing?: string;
  cursor?: string;
  position?: string;
  zIndex?: string;
}

export type ElementOverlayOptions = {
  className?: string;
  style?: ElementOverlayStyleOptions;
};

export const getElementBounds = (el: HTMLElement): BoundingBox => {
  const rect = el.getBoundingClientRect();
  return {
    height: el.offsetHeight,
    width: el.offsetWidth,
    x: window.pageXOffset + rect.left,
    y: window.pageYOffset + rect.top,
  };
};
