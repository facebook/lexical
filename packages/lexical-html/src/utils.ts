/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

export function cache2Get<A, B, C>(
  cache: Map<A, Map<B, C>>,
  a: A,
  b: B,
): undefined | C {
  const nested = cache.get(a);
  if (nested !== undefined) {
    return nested.get(b);
  }
}

export function cache2Set<A, B, C>(
  cache: Map<A, Map<B, C>>,
  a: A,
  b: B,
  value: C,
): void {
  let nested = cache.get(a);
  if (nested === undefined) {
    nested = new Map<B, C>();
    cache.set(a, nested);
  }
  nested.set(b, value);
}
