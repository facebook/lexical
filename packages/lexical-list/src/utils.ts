/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import type {LexicalNode} from 'lexical';

import invariant from 'shared/invariant';

import {
  $createListItemNode,
  $isListItemNode,
  $isListNode,
  ListItemNode,
  ListNode,
} from './';

/**
 * Checks the depth of listNode from the root node.
 * @param listNode - The ListNode to be checked.
 * @returns The depth of the ListNode.
 */
export function $getListDepth(listNode: ListNode): number {
  let depth = 1;
  let parent = listNode.getParent();

  while (parent != null) {
    if ($isListItemNode(parent)) {
      const parentList = parent.getParent();

      if ($isListNode(parentList)) {
        depth++;
        parent = parentList.getParent();
        continue;
      }
      invariant(false, 'A ListItemNode must have a ListNode for a parent.');
    }

    return depth;
  }

  return depth;
}

/**
 * Finds the nearest ancestral ListNode and returns it, throws an invariant if listItem is not a ListItemNode.
 * @param listItem - The node to be checked.
 * @returns The ListNode found.
 */
export function $getTopListNode(listItem: LexicalNode): ListNode {
  let list = listItem.getParent<ListNode>();

  if (!$isListNode(list)) {
    invariant(false, 'A ListItemNode must have a ListNode for a parent.');
  }

  let parent: ListNode | null = list;

  while (parent !== null) {
    parent = parent.getParent();

    if ($isListNode(parent)) {
      list = parent;
    }
  }

  return list;
}

/**
 * Checks if listItem has no child ListNodes and has no ListItemNode ancestors with siblings.
 * @param listItem - the ListItemNode to be checked.
 * @returns true if listItem has no child ListNode and no ListItemNode ancestors with siblings, false otherwise.
 */
export function $isLastItemInList(listItem: ListItemNode): boolean {
  let isLast = true;
  const firstChild = listItem.getFirstChild();

  if ($isListNode(firstChild)) {
    return false;
  }
  let parent: ListItemNode | null = listItem;

  while (parent !== null) {
    if ($isListItemNode(parent)) {
      if (parent.getNextSiblings().length > 0) {
        isLast = false;
      }
    }

    parent = parent.getParent();
  }

  return isLast;
}

/**
 * A recursive Depth-First Search (Postorder Traversal) that finds all of a node's children
 * that are of type ListItemNode and returns them in an array.
 * @param node - The ListNode to start the search.
 * @returns An array containing all nodes of type ListItemNode found.
 */
// This should probably be $getAllChildrenOfType
export function $getAllListItems(node: ListNode): Array<ListItemNode> {
  let listItemNodes: Array<ListItemNode> = [];
  const listChildren: Array<ListItemNode> = node
    .getChildren()
    .filter($isListItemNode);

  for (let i = 0; i < listChildren.length; i++) {
    const listItemNode = listChildren[i];
    const firstChild = listItemNode.getFirstChild();

    if ($isListNode(firstChild)) {
      listItemNodes = listItemNodes.concat($getAllListItems(firstChild));
    } else {
      listItemNodes.push(listItemNode);
    }
  }

  return listItemNodes;
}

/**
 * Checks to see if the passed node is a ListItemNode and has a ListNode as a child.
 * @param node - The node to be checked.
 * @returns true if the node is a ListItemNode and has a ListNode child, false otherwise.
 */
export function isNestedListNode(
  node: LexicalNode | null | undefined,
): boolean {
  return $isListItemNode(node) && $isListNode(node.getFirstChild());
}

/**
 * Traverses up the tree and returns the first ListItemNode found.
 * @param node - Node to start the search.
 * @returns The first ListItemNode found, or null if none exist.
 */
// TODO: rewrite with $findMatchingParent or *nodeOfType
export function findNearestListItemNode(
  node: LexicalNode,
): ListItemNode | null {
  let currentNode: LexicalNode | null = node;

  while (currentNode !== null) {
    if ($isListItemNode(currentNode)) {
      return currentNode;
    }
    currentNode = currentNode.getParent();
  }

  return null;
}

/**
 * Takes a deeply nested ListNode or ListItemNode and traverses up the branch to delete the first
 * ancestral ListNode (which could be the root ListNode) or ListItemNode with siblings, essentially
 * bringing the deeply nested node up the branch once. Would remove sublist if it has siblings.
 * Should not break ListItem -> List -> ListItem chain as empty List/ItemNodes should be removed on .remove().
 * @param sublist - The nested ListNode or ListItemNode to be brought up the branch.
 */
export function $removeHighestEmptyListParent(
  sublist: ListItemNode | ListNode,
) {
  // Nodes may be repeatedly indented, to create deeply nested lists that each
  // contain just one bullet.
  // Our goal is to remove these (empty) deeply nested lists. The easiest
  // way to do that is crawl back up the tree until we find a node that has siblings
  // (e.g. is actually part of the list contents) and delete that, or delete
  // the root of the list (if no list nodes have siblings.)
  let emptyListPtr = sublist;

  while (
    emptyListPtr.getNextSibling() == null &&
    emptyListPtr.getPreviousSibling() == null
  ) {
    const parent = emptyListPtr.getParent<ListItemNode | ListNode>();

    if (
      parent == null ||
      !($isListItemNode(emptyListPtr) || $isListNode(emptyListPtr))
    ) {
      break;
    }

    emptyListPtr = parent;
  }

  emptyListPtr.remove();
}

/**
 * Wraps a node into a ListItemNode.
 * @param node - The node to be wrapped into a ListItemNode
 * @returns The ListItemNode which the passed node is wrapped in.
 */
export function wrapInListItem(node: LexicalNode): ListItemNode {
  const listItemWrapper = $createListItemNode();
  return listItemWrapper.append(node);
}
