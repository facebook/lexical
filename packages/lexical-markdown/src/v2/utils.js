/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow strict
 */

import type {
  ElementTransformer,
  TextFormatTransformer,
  TextMatchTransformer,
  Transformer,
} from '../../flow/LexicalMarkdown';

export function indexBy<T>(
  list: Array<T>,
  callback: (T) => string,
): $ReadOnly<{[string]: Array<T>}> {
  const index = {};
  for (const item of list) {
    const key = callback(item);
    if (index[key]) {
      index[key].push(item);
    } else {
      index[key] = [item];
    }
  }
  return index;
}

export function transformersByType(
  transformers: Array<Transformer>,
): $ReadOnly<{
  element: Array<ElementTransformer>,
  textFormat: Array<TextFormatTransformer>,
  textMatch: Array<TextMatchTransformer>,
}> {
  const byType = indexBy(transformers, (t) => t.type);
  return {
    // $FlowFixMe
    element: byType.element,
    // $FlowFixMe
    textFormat: byType['text-format'],
    // $FlowFixMe
    textMatch: byType['text-match'],
  };
}
