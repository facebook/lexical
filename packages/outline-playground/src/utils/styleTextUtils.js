/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow strict-local
 */

import type {Selection, TextNode} from 'outline';

import {isTextNode} from 'outline';
import {getStyleObjectFromCSS} from 'outline/SelectionHelpers';

export function getSelectionFontSize(
  selection: Selection,
  defaultValue: string = '',
): string {
  let fontSize = null;
  const nodes = selection.getNodes();
  for (let i = 0; i < nodes.length; i++) {
    const node = nodes[i];
    if (isTextNode(node)) {
      const nodeFontSize = getNodeFontSize(node, defaultValue);
      if (fontSize === null) {
        fontSize = nodeFontSize;
      } else if (fontSize !== nodeFontSize) {
        // multiple text nodes are in the selection and they don't all
        // have the same font size.
        fontSize = '';
        break;
      }
    }
  }
  return fontSize || '';
}

function getNodeFontSize(node: TextNode, defaultValue: string): string {
  const css = node.getStyle();
  const styleObject = getStyleObjectFromCSS(css);
  return styleObject ? styleObject['font-size'] : defaultValue;
}
