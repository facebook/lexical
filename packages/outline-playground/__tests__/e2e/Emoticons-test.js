/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import {initializeE2E, assertHTML, assertSelection, repeat} from '../utils';
import {moveToLineBeginning, moveToLineEnd} from '../keyboardShortcuts';

describe('Emoticons', () => {
  initializeE2E((e2e) => {
    it(`Can handle a single emoticon`, async () => {
      const {page} = e2e;

      await page.focus('div.editor');
      await page.keyboard.type('This is an emoji :)');
      await assertHTML(
        page,
        '<p class="editor-paragraph" dir="ltr"><span data-outline-text="true">This is an emoji </span><span class="emoji happysmile" data-outline-text="true">🙂</span></p>',
      );
      await assertSelection(page, {
        anchorPath: [0, 1, 0],
        anchorOffset: 2,
        focusPath: [0, 1, 0],
        focusOffset: 2,
      });

      await page.keyboard.press('Backspace');
      await assertHTML(
        page,
        '<p class="editor-paragraph" dir="ltr"><span data-outline-text="true">This is an emoji </span></p>',
      );
      await assertSelection(page, {
        anchorPath: [0, 0, 0],
        anchorOffset: 17,
        focusPath: [0, 0, 0],
        focusOffset: 17,
      });

      await page.keyboard.type(':)');
      await page.keyboard.press('ArrowLeft');
      await assertSelection(page, {
        anchorPath: [0, 0, 0],
        anchorOffset: 17,
        focusPath: [0, 0, 0],
        focusOffset: 17,
      });

      await page.keyboard.press('ArrowRight');
      await assertSelection(page, {
        anchorPath: [0, 1, 0],
        anchorOffset: 2,
        focusPath: [0, 1, 0],
        focusOffset: 2,
      });

      await page.keyboard.press('ArrowLeft');
      await page.keyboard.press('Delete');
      await assertHTML(
        page,
        '<p class="editor-paragraph" dir="ltr"><span data-outline-text="true">This is an emoji </span></p>',
      );
      await assertSelection(page, {
        anchorPath: [0, 0, 0],
        anchorOffset: 17,
        focusPath: [0, 0, 0],
        focusOffset: 17,
      });
    });

    it(`Can enter mutliple emoticons`, async () => {
      const {isRichText, page} = e2e;

      await page.focus('div.editor');
      await page.keyboard.type(':) :) <3 :(');
      await assertHTML(
        page,
        '<p class="editor-paragraph"><span class="emoji happysmile" data-outline-text="true">🙂</span><span data-outline-text="true"> </span><span class="emoji happysmile" data-outline-text="true">🙂</span><span data-outline-text="true"> </span><span class="emoji heart" data-outline-text="true">❤</span><span data-outline-text="true"> </span><span class="emoji unhappysmile" data-outline-text="true">🙁</span></p>',
      );
      await assertSelection(page, {
        anchorPath: [0, 6, 0],
        anchorOffset: 2,
        focusPath: [0, 6, 0],
        focusOffset: 2,
      });

      await page.keyboard.down('Shift');
      await page.keyboard.press('Enter');
      await page.keyboard.up('Shift');
      await assertHTML(
        page,
        '<p class="editor-paragraph"><span class="emoji happysmile" data-outline-text="true">🙂</span><span data-outline-text="true"> </span><span class="emoji happysmile" data-outline-text="true">🙂</span><span data-outline-text="true"> </span><span class="emoji heart" data-outline-text="true">❤</span><span data-outline-text="true"> </span><span class="emoji unhappysmile" data-outline-text="true">🙁</span><br><br></p>',
      );
      await assertSelection(page, {
        anchorPath: [0],
        anchorOffset: 8,
        focusPath: [0],
        focusOffset: 8,
      });

      await page.keyboard.type(':) :) <3 :(');
      await assertHTML(
        page,
        '<p class="editor-paragraph"><span class="emoji happysmile" data-outline-text="true">🙂</span><span data-outline-text="true"> </span><span class="emoji happysmile" data-outline-text="true">🙂</span><span data-outline-text="true"> </span><span class="emoji heart" data-outline-text="true">❤</span><span data-outline-text="true"> </span><span class="emoji unhappysmile" data-outline-text="true">🙁</span><br><span class="emoji happysmile" data-outline-text="true">🙂</span><span data-outline-text="true"> </span><span class="emoji happysmile" data-outline-text="true">🙂</span><span data-outline-text="true"> </span><span class="emoji heart" data-outline-text="true">❤</span><span data-outline-text="true"> </span><span class="emoji unhappysmile" data-outline-text="true">🙁</span></p>',
      );
      await assertSelection(page, {
        anchorPath: [0, 14, 0],
        anchorOffset: 2,
        focusPath: [0, 14, 0],
        focusOffset: 2,
      });

      await page.keyboard.press('Enter');
      if (isRichText) {
        await assertHTML(
          page,
          '<p class="editor-paragraph"><span class="emoji happysmile" data-outline-text="true">🙂</span><span data-outline-text="true"> </span><span class="emoji happysmile" data-outline-text="true">🙂</span><span data-outline-text="true"> </span><span class="emoji heart" data-outline-text="true">❤</span><span data-outline-text="true"> </span><span class="emoji unhappysmile" data-outline-text="true">🙁</span><br><span class="emoji happysmile" data-outline-text="true">🙂</span><span data-outline-text="true"> </span><span class="emoji happysmile" data-outline-text="true">🙂</span><span data-outline-text="true"> </span><span class="emoji heart" data-outline-text="true">❤</span><span data-outline-text="true"> </span><span class="emoji unhappysmile" data-outline-text="true">🙁</span></p><p class="editor-paragraph"><br></p>',
        );
        await assertSelection(page, {
          anchorPath: [1],
          anchorOffset: 0,
          focusPath: [1],
          focusOffset: 0,
        });
      } else {
        await assertHTML(
          page,
          '<p class="editor-paragraph"><span class="emoji happysmile" data-outline-text="true">🙂</span><span data-outline-text="true"> </span><span class="emoji happysmile" data-outline-text="true">🙂</span><span data-outline-text="true"> </span><span class="emoji heart" data-outline-text="true">❤</span><span data-outline-text="true"> </span><span class="emoji unhappysmile" data-outline-text="true">🙁</span><br><span class="emoji happysmile" data-outline-text="true">🙂</span><span data-outline-text="true"> </span><span class="emoji happysmile" data-outline-text="true">🙂</span><span data-outline-text="true"> </span><span class="emoji heart" data-outline-text="true">❤</span><span data-outline-text="true"> </span><span class="emoji unhappysmile" data-outline-text="true">🙁</span><br><br></p>',
        );
        await assertSelection(page, {
          anchorPath: [0],
          anchorOffset: 16,
          focusPath: [0],
          focusOffset: 16,
        });
      }

      await page.keyboard.type(':) :) <3 :(');
      if (isRichText) {
        await assertHTML(
          page,
          '<p class="editor-paragraph"><span class="emoji happysmile" data-outline-text="true">🙂</span><span data-outline-text="true"> </span><span class="emoji happysmile" data-outline-text="true">🙂</span><span data-outline-text="true"> </span><span class="emoji heart" data-outline-text="true">❤</span><span data-outline-text="true"> </span><span class="emoji unhappysmile" data-outline-text="true">🙁</span><br><span class="emoji happysmile" data-outline-text="true">🙂</span><span data-outline-text="true"> </span><span class="emoji happysmile" data-outline-text="true">🙂</span><span data-outline-text="true"> </span><span class="emoji heart" data-outline-text="true">❤</span><span data-outline-text="true"> </span><span class="emoji unhappysmile" data-outline-text="true">🙁</span></p><p class="editor-paragraph"><span class="emoji happysmile" data-outline-text="true">🙂</span><span data-outline-text="true"> </span><span class="emoji happysmile" data-outline-text="true">🙂</span><span data-outline-text="true"> </span><span class="emoji heart" data-outline-text="true">❤</span><span data-outline-text="true"> </span><span class="emoji unhappysmile" data-outline-text="true">🙁</span></p>',
        );
        await assertSelection(page, {
          anchorPath: [1, 6, 0],
          anchorOffset: 2,
          focusPath: [1, 6, 0],
          focusOffset: 2,
        });
      } else {
        await assertHTML(
          page,
          '<p class="editor-paragraph"><span class="emoji happysmile" data-outline-text="true">🙂</span><span data-outline-text="true"> </span><span class="emoji happysmile" data-outline-text="true">🙂</span><span data-outline-text="true"> </span><span class="emoji heart" data-outline-text="true">❤</span><span data-outline-text="true"> </span><span class="emoji unhappysmile" data-outline-text="true">🙁</span><br><span class="emoji happysmile" data-outline-text="true">🙂</span><span data-outline-text="true"> </span><span class="emoji happysmile" data-outline-text="true">🙂</span><span data-outline-text="true"> </span><span class="emoji heart" data-outline-text="true">❤</span><span data-outline-text="true"> </span><span class="emoji unhappysmile" data-outline-text="true">🙁</span><br><span class="emoji happysmile" data-outline-text="true">🙂</span><span data-outline-text="true"> </span><span class="emoji happysmile" data-outline-text="true">🙂</span><span data-outline-text="true"> </span><span class="emoji heart" data-outline-text="true">❤</span><span data-outline-text="true"> </span><span class="emoji unhappysmile" data-outline-text="true">🙁</span></p>',
        );
        await assertSelection(page, {
          anchorPath: [0, 22, 0],
          anchorOffset: 2,
          focusPath: [0, 22, 0],
          focusOffset: 2,
        });
      }

      await moveToLineBeginning(page);
      // This should not crash on a deletion on an immutable node
      await page.keyboard.press('Backspace');
      await moveToLineEnd(page);

      await repeat(22, async () => await page.keyboard.press('Backspace'));
      await assertHTML(
        page,
        '<p class="editor-paragraph"><span data-outline-text="true"><br></span></p>',
      );
      await assertSelection(page, {
        anchorPath: [0, 0, 0],
        anchorOffset: 0,
        focusPath: [0, 0, 0],
        focusOffset: 0,
      });

      await page.keyboard.type(':):):):):)');
      await page.keyboard.press('ArrowLeft');
      await assertSelection(page, {
        anchorPath: [0, 3, 0],
        anchorOffset: 2,
        focusPath: [0, 3, 0],
        focusOffset: 2,
      });

      await page.keyboard.press('ArrowLeft');
      await assertSelection(page, {
        anchorPath: [0, 2, 0],
        anchorOffset: 2,
        focusPath: [0, 2, 0],
        focusOffset: 2,
      });

      await page.keyboard.press('ArrowLeft');
      await assertSelection(page, {
        anchorPath: [0, 1, 0],
        anchorOffset: 2,
        focusPath: [0, 1, 0],
        focusOffset: 2,
      });

      await page.keyboard.press('ArrowLeft');
      await assertSelection(page, {
        anchorPath: [0, 0, 0],
        anchorOffset: 2,
        focusPath: [0, 0, 0],
        focusOffset: 2,
      });

      await page.keyboard.press('ArrowLeft');
      await assertSelection(page, {
        anchorPath: [0, 0, 0],
        anchorOffset: 0,
        focusPath: [0, 0, 0],
        focusOffset: 0,
      });

      await page.keyboard.type('Hey');
      await assertHTML(
        page,
        '<p class="editor-paragraph" dir="ltr"><span data-outline-text="true">Hey</span><span class="emoji happysmile" data-outline-text="true">🙂</span><span class="emoji happysmile" data-outline-text="true">🙂</span><span class="emoji happysmile" data-outline-text="true">🙂</span><span class="emoji happysmile" data-outline-text="true">🙂</span><span class="emoji happysmile" data-outline-text="true">🙂</span></p>',
      );
      await assertSelection(page, {
        anchorPath: [0, 0, 0],
        anchorOffset: 3,
        focusPath: [0, 0, 0],
        focusOffset: 3,
      });
    });
  });
});
