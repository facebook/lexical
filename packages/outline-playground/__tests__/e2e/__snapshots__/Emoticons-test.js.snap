// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Emoticons Can enter mutliple emoticons 1`] = `
<p
  class="editor-paragraph"
  dir="ltr"
>
  <span>
    ﻿
  </span>
  <span
    class="emoji happysmile"
    contenteditable="false"
  >
    🙂
  </span>
  <span>
     
  </span>
  <span
    class="emoji happysmile"
    contenteditable="false"
  >
    🙂
  </span>
  <span>
     
  </span>
  <span
    class="emoji heart"
    contenteditable="false"
  >
    ❤
  </span>
  <span>
     
  </span>
  <span
    class="emoji unhappysmile"
    contenteditable="false"
  >
    🙁
  </span>
  <span>
    ﻿
  </span>
</p>
`;

exports[`Emoticons Can enter mutliple emoticons 2`] = `
<p
  class="editor-paragraph"
  dir="ltr"
>
  <span>
    ﻿
  </span>
  <span
    class="emoji happysmile"
    contenteditable="false"
  >
    🙂
  </span>
  <span>
     
  </span>
  <span
    class="emoji happysmile"
    contenteditable="false"
  >
    🙂
  </span>
  <span>
     
  </span>
  <span
    class="emoji heart"
    contenteditable="false"
  >
    ❤
  </span>
  <span>
     
  </span>
  <span
    class="emoji unhappysmile"
    contenteditable="false"
  >
    🙁
  </span>
  <span>
    ﻿
  </span>
  <br />
  <span>
    ﻿
  </span>
</p>
`;

exports[`Emoticons Can enter mutliple emoticons 3`] = `
<p
  class="editor-paragraph"
  dir="ltr"
>
  <span>
    ﻿
  </span>
  <span
    class="emoji happysmile"
    contenteditable="false"
  >
    🙂
  </span>
  <span>
     
  </span>
  <span
    class="emoji happysmile"
    contenteditable="false"
  >
    🙂
  </span>
  <span>
     
  </span>
  <span
    class="emoji heart"
    contenteditable="false"
  >
    ❤
  </span>
  <span>
     
  </span>
  <span
    class="emoji unhappysmile"
    contenteditable="false"
  >
    🙁
  </span>
  <span>
    ﻿
  </span>
  <br />
  <span>
    ﻿
  </span>
  <span
    class="emoji happysmile"
    contenteditable="false"
  >
    🙂
  </span>
  <span>
     
  </span>
  <span
    class="emoji happysmile"
    contenteditable="false"
  >
    🙂
  </span>
  <span>
     
  </span>
  <span
    class="emoji heart"
    contenteditable="false"
  >
    ❤
  </span>
  <span>
     
  </span>
  <span
    class="emoji unhappysmile"
    contenteditable="false"
  >
    🙁
  </span>
  <span>
    ﻿
  </span>
</p>
`;

exports[`Emoticons Can enter mutliple emoticons 4`] = `
<p
  class="editor-paragraph"
  dir="ltr"
>
  <span>
    ﻿
  </span>
  <span
    class="emoji happysmile"
    contenteditable="false"
  >
    🙂
  </span>
  <span>
     
  </span>
  <span
    class="emoji happysmile"
    contenteditable="false"
  >
    🙂
  </span>
  <span>
     
  </span>
  <span
    class="emoji heart"
    contenteditable="false"
  >
    ❤
  </span>
  <span>
     
  </span>
  <span
    class="emoji unhappysmile"
    contenteditable="false"
  >
    🙁
  </span>
  <span>
    ﻿
  </span>
  <br />
  <span>
    ﻿
  </span>
  <span
    class="emoji happysmile"
    contenteditable="false"
  >
    🙂
  </span>
  <span>
     
  </span>
  <span
    class="emoji happysmile"
    contenteditable="false"
  >
    🙂
  </span>
  <span>
     
  </span>
  <span
    class="emoji heart"
    contenteditable="false"
  >
    ❤
  </span>
  <span>
     
  </span>
  <span
    class="emoji unhappysmile"
    contenteditable="false"
  >
    🙁
  </span>
  <span>
    ﻿
  </span>
</p>
`;

exports[`Emoticons Can enter mutliple emoticons 5`] = `
<p
  class="editor-paragraph"
  dir="ltr"
>
  <span>
    ﻿
  </span>
  <span
    class="emoji happysmile"
    contenteditable="false"
  >
    🙂
  </span>
  <span>
     
  </span>
  <span
    class="emoji happysmile"
    contenteditable="false"
  >
    🙂
  </span>
  <span>
     
  </span>
  <span
    class="emoji heart"
    contenteditable="false"
  >
    ❤
  </span>
  <span>
     
  </span>
  <span
    class="emoji unhappysmile"
    contenteditable="false"
  >
    🙁
  </span>
  <span>
    ﻿
  </span>
  <br />
  <span>
    ﻿
  </span>
  <span
    class="emoji happysmile"
    contenteditable="false"
  >
    🙂
  </span>
  <span>
     
  </span>
  <span
    class="emoji happysmile"
    contenteditable="false"
  >
    🙂
  </span>
  <span>
     
  </span>
  <span
    class="emoji heart"
    contenteditable="false"
  >
    ❤
  </span>
  <span>
     
  </span>
  <span
    class="emoji unhappysmile"
    contenteditable="false"
  >
    🙁
  </span>
  <span>
    ﻿
  </span>
</p>
`;

exports[`Emoticons Can enter mutliple emoticons 6`] = `
<p
  class="editor-paragraph"
>
  <span>
    ﻿
  </span>
</p>
`;

exports[`Emoticons Can handle a single emoticon 1`] = `
<p
  class="editor-paragraph"
  dir="ltr"
>
  <span>
    This is an emoji 
  </span>
  <span
    class="emoji happysmile"
    contenteditable="false"
  >
    🙂
  </span>
  <span>
    ﻿
  </span>
</p>
`;

exports[`Emoticons Can handle a single emoticon 2`] = `
<p
  class="editor-paragraph"
  dir="ltr"
>
  <span>
    This is an emoji 
  </span>
</p>
`;

exports[`Emoticons Can handle a single emoticon 3`] = `
<p
  class="editor-paragraph"
  dir="ltr"
>
  <span>
    This is an emoji 
  </span>
</p>
`;
